packages/@expo/cli/src/utils/codesigning.ts

```diff
@@ -0,0 +1,298 @@
+ * - <developer's expo-updates keyid> indicates that it should sign with the configured certificate. See {@link getProjectCodeSigningCertificateAsync}
+ */
+export async function getCodeSigningInfoAsync(
+  projectRoot: string,
+  exp: ExpoConfig,
+  expectSignatureHeader: string | null,
+  privateKeyPath: string | undefined
+): Promise<CodeSigningInfo | null> {
+  if (!expectSignatureHeader) {
+    return null;
+  }
+
+  let parsedExpectSignature: Dictionary;
+  try {
+    parsedExpectSignature = parseDictionary(expectSignatureHeader);
+  } catch {
+    throw new Error('Invalid value for expo-expect-signature header');
+  }
+
+  const expectedKeyIdOuter = parsedExpectSignature.get('keyid');
+  if (!expectedKeyIdOuter) {
+    throw new Error('keyid not present in expo-expect-signature header');
+  }
+
+  const expectedKeyId = expectedKeyIdOuter[0];
+  if (typeof expectedKeyId !== 'string') {
+    throw new Error(`Invalid value for keyid in expo-expect-signature header: ${expectedKeyId}`);
+  }
+
+  let expectedAlg: string | null = null;
+  const expectedAlgOuter = parsedExpectSignature.get('alg');
+  if (expectedAlgOuter) {
+    const expectedAlgTemp = expectedAlgOuter[0];
+    if (typeof expectedAlgTemp !== 'string') {
+      throw new Error('Invalid value for alg in expo-expect-signature header');
+    }
+    expectedAlg = expectedAlgTemp;
+  }
+
+  if (expectedKeyId === 'expo-root') {
+    return await getExpoRootDevelopmentCodeSigningInfoAsync(projectRoot, exp);
+  } else if (expectedKeyId === 'expo-go') {
+    throw new Error('Invalid certificate requested: cannot sign with embedded keyid=expo-go key');
+  } else {
+    return await getProjectCodeSigningCertificateAsync(
+      exp,
+      privateKeyPath,
+      expectedKeyId,
+      expectedAlg
+    );
+  }
+}
+
+/**
+ * Get a development code signing certificate for the expo-root -> expo-go -> (development certificate) certificate chain.
+ * This requires the user be logged in and online, otherwise try to use the cached development certificate.
+ */
+async function getExpoRootDevelopmentCodeSigningInfoAsync(
+  projectRoot: string,
+  exp: ExpoConfig
+): Promise<CodeSigningInfo | null> {
+  const easProjectId = exp.extra?.eas?.projectId;
+  // can't check for scope key validity since scope key is derived on the server from projectId and we may be offline.
+  // we rely upon the client certificate check to validate the scope key
+  if (!easProjectId) {
+    Log.warn('No project ID specified in app.json, unable to sign manifest');
+    return null;
+  }
+
+  // 1. If online, ensure logged in, generate PK, CSR, fetch and cache cert chain for projectId
+  //    (overwriting existing dev cert in case projectId changed or it has expired)
+  if (!APISettings.isOffline) {
+    return await fetchAndCacheNewDevelopmentCodeSigningInfoAsync(projectRoot, easProjectId);
```

Comment:
Even if running in "online" mode, if there's a network error or server error when fetching the certificates, it would be nice to be able to fall back to the local copies if they haven't expired. That logic probably belongs in `fetchAndCacheNewDevelopmentCodeSigningInfoAsync` but maybe part of the logic belongs at this layer.

Reply:
This is a good idea, but one that I think will require quite a bit more code to be able to discern between different types of errors. I'll follow-up in a separate PR: https://linear.app/expo/issue/ENG-4529/serve-cached-dev-certificate-if-expo-servers-are-down